first commit for chordShapes function

grimmdude opened the pull request:
Hi @danigb ,

Here's my first go for the `chordShapes` function.  I noticed the `dist/tonal.js` was modified when I ran the build commands, hope that's not a problem?   Anyway, I think more elegance could be had in this function, but I figured I'd send it your way and see what you thought.

The basic idea is you can supply the tuning, the chord tones, the reach of frets, and it will output arrays of notes starting on each fret that can be gripped within the fret range supplied.  The notes are indexed by string.  For instance this would be an open E chord:

`[ 'E2', 'B2', 'E3', 'G#3', 'B3', 'E4' ]`

In cases where there could be multiple notes fretted per string that element will be an array.  For instance the first result for Emaj7:

`[ 'E2', 'B2', [ 'D#3', 'E3' ], 'G#3', 'B3', 'E4' ]`

#8

danigb commented:
Hi Garret,

Thanks for the PR! 👍  

My first thoughts are that I don't know If I understand clearly the purpose of your function :-D. As I understand, it returns you the notes of the chord/scale for a fret range. It's like asking: how I can play this chord in this position? If so, I think it would be nice to know, not only the string note, but also the fret number to produce... But maybe I'm totally wrong.

Anyway, I will take a look at the code, and give you better feedback.

Thanks a lot!

grimmdude commented:
Right, the simple point is what you said: 'returns you the notes of the chord/scale for a fret range'.  I initially thought including the fret number would be good, but in the interest of keeping the data structure simple I opted to leave it out since it could technically be derived from the data given and tuning.  On that note, having a function that returns the fret number of a specific string/note could be nice to have too.   Thanks for checking it out!

danigb commented:
Aja, I understand. But for me, seems more natural that your function returns the fret numbers (instead of the notes). It's more a "figure", and I think it's easy to get the notes from the frets numbers than the opposite... What do you think?

danigb commented:
Also, some other test would be really appreciated. Showing things like an empty range, or very narrow one...

grimmdude commented:
Yea I agree, fret numbers would make more sense in this case.  And yes, I can add more tests.  Thanks,

-G

danigb commented:
Great! Let's do this PR 👍

grimmdude commented:
Cool, I've changed this to return fret numbers and added a couple of tests.  I don't do much test driven stuff myself, so let me know if you would like to see more.  

I think the basic functionality of this is working great.  Do you think shapes with only a single available fret be omitted?  Currently it's returning those.

Also, I noticed `tonalArray.compact()` removes `0` values as well as `null` fyi.

danigb commented:
Hello Garret. Thanks for the updates! I think you're building a great functionality. But before merge, I have some issues with the code that I'll comment inline. Take them as suggestions, and we discuss about them.

And good catch with the `compact` function.  👍 

Thanks again!

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+ * @param {Integer} last - the last fret number
+ * @return {Array} An array of arrays, one for each possible shape.  Index is string index [ 'E2', 'B2', 'E3', 'G#3', 'B3', 'E4' ]
+ */
+export function chordShapes (tuning, notes, first, last, span) {
```
I think we could find a better name than `chordShapes`. Something more concrete and descriptive. What about: chordPositions? I don't have the perfect name, but we should try to find it (tonal has a quite big API and it's important to grasp easily what is all about)

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+ * @param {String|Array} tuning - the tuning name or notes
+ * @param {Array} notes - an array of chord notes
+ * @param {Integer} first - the first fret number
+ * @param {Integer} last - the last fret number
```
Missing the `span` parameter

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+ * @param {Array} notes - an array of chord notes
+ * @param {Integer} first - the first fret number
+ * @param {Integer} last - the last fret number
+ * @return {Array} An array of arrays, one for each possible shape.  Index is string index [ 'E2', 'B2', 'E3', 'G#3', 'B3', 'E4' ]
```
The return example is using notes instead of fret numbers.

grimmdude replied:
Fixed

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+ */
+export function chordShapes (tuning, notes, first, last, span) {
+  var chordNotes = scale(tuning, notes, first, last)
+  var chordGroups = []
```
What is a `chordGroup`? Can we find a better name?

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+        return v3 !== null ? i2 + i3  : null
+      })
+    })
+  })
```
It's not clear what this double loop does (working with nested arrays is always hard...). I suggest you to have more descriptive variable names: for example ... v?

grimmdude replied:
Changed the var names to be more descriptive

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+      // Strip null values
+      var frets = v2.filter(function(v3) {
+        return v3 !== null
+      })
```
compact bug was a very nice caught. Now it's fixed on master, so I prefer to use it (and then remove the comment: we use named functions to make the code self descriptive about the intentions... sorry if I'm too pedantic).

grimmdude replied:
I merged your upstream master with this fix but it still seems to be stripping `0` values.  `console.log(compact([null,0,1]));` returns `[1]`

grimmdude replied:
Although, I see that the tests for `tonal-array` pass so I must be doing something wrong on my end.

danigb replied:
I know what the problem is: tonal-fretboard depends on tonal but I din't release a tonal package, only the tonal-array package.
For me the best solution is to change tonal-fretboard dependencies, so it uses only the libraries from tonal it needs...

grimmdude replied:
Ah ok, that makes sense.  I think I'll leave that dependency change to you

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+      }
+    })
+  })
+
```
Again this is hard to understand. What is frets.toString()? Maybe use a named variable to describe what it is, and then simplify the code: `...push(xx ? xx : null)`

danigb commented on modules/fretboard/index.js:
```diff
@@ -71,3 +71,50 @@ export function scale (tuning, scale, first, last) {
+
+  // Remove null and neighboring duplicate arrays
+  return chordGroups.filter(function(v, i) {
+    if (!v.join('').toString()) return false;
```
The same again... what does v.join('').toString() ... maybe create a function called isFretsEmpty(v) or similar...

grimmdude replied:
This checks if the array is full of `null` values.  What about adding a method to `tonal-array` like `isNull(array)`?

danigb replied:
I would rewrite that snippet like: `...filter(function(v) { return compact(v).length })` 

For me it's clear enough... What do you think?

grimmdude replied:
Yep, that will work fine.  See my other note about `compact()` still stripping `0` values though.

-G

danigb commented:
Hi Garret... I made some comments to your code. Don't be intimidated by the number... I spent lot of time trying to make tonal code readable and I would like to have the same attention to your code. Anyway, I'm so excited with your functionality that I'm planning to make a fretboard drawer: https://github.com/danigb/fretboard  ... Suggestions and collaboration is welcomed ;-)

grimmdude commented:
Hi @danigb,

I definitely appreciate the feedback, I was hoping you would review it.  Thanks, I'll go through your notes and make necessary changes.

As far as the name, I was thinking it could almost be named `scalePositions` because any range of notes can be passed to it.

danigb commented:
Hi @grimmdude 

Happy you didn't scare with the comments. 

Regarding the name I'm thinking in change the names of some fretboard module functions. In fact, I was thinking in merge `build` and `scale` function into one called `notes` or `notesByString`. In that new function, if you pass a `filter` parameter, the notes are filtered (like the current `scales`) and if not, you get the behaviour of the current `build` function.

With the same logic, there would be a function called `positions` or `positionsByString` that returns the positions optionally filtered by some notes or pitch classes.

I think your function is an extension of this `positions` function, because the span parameter... Isn't it?

grimmdude commented:
That makes sense.  Yes by that logic this function would extend that `positions` function by limiting results by `span`.  Now that we have the objective of this function, I think we can split it off into smaller, reusable parts.

I've made the changes per your notes, with exception of the code that requires `compare`.  Do you want to make those dependency changes in your master and I can pull it down?

-G

danigb commented:
Hi @grimmdude, nice work!! Thanks. For me it's time to merge your changes into master and continue there... what do you think?

grimmdude commented:
Awesome thanks. I've pulled down your changes so I'm all up to date.  I'll continue thinking of ways to optimize this function, as well as other functionality we could add to the module.

-G

Euthymious commented:
Hi, @danigb and @grimmdude,
I see there's no recent activity in this thread, but may I be so bold to revive it?

I was trying to implement a chord diagram generator (yes, yet another one), just as a practical exercise in javascript, and I came up with a dirty but working solution. Then I looked for other implementations -- to see, how should it be done 'the right way' -- and I found your project. While not being a programmer at all, I dare to note, that you're doing something whrong with those chord shapes.

First, the output of your chordShapes function seems counterintuitive to me: since it can return multiple notes fretted per string, I'd rather call it arpeggio patterns. To get real chord diagrams from that one should post-process it somehow, and that seems to be a non-trivial task.

I looked what @grimmdude does here: https://github.com/grimmdude/musictheorysite.com/blob/master/assets/js/guitar-chords/index.js
As far as I can see, just takes a first element of an array and that gives weird results.

Take a look, for example, at Eadd9 chord. The first shape you get is 0-2-2-1-0-0 which is just plain E with no added 9th. Then you come with valid fingering 2-2-2-1-x-2 but omit  0-2-2-1-0-2 (just as valid), and later on get 7-7-6-9-7-7, which is practically infingerable.

I would propose another variant, producing fingerings as an array with exactly one fretting position per string ('x' for muted string): 
https://runkit.com/euthymious/chordshapesalt-2
This one is still far from perfect, as the fingerings are checked just for completeness (i.e. every note of the chord is included), not for physical 'fingerability', but i guess additional checks can be added easily.

Thank you for your attention.
